src/storage/v2/durability/snapshot.cpp

```diff
@@ -157,17 +187,378 @@ SnapshotInfo ReadSnapshotInfo(const std::filesystem::path &path) {
+      }
+    }
+
+    // Recover out edges.
+    {
+      spdlog::trace("Recovering outbound edges for vertex {}.", vertex.gid.AsUint());
+      auto out_size = snapshot.ReadUint();
+      if (!out_size) throw RecoveryFailure("Invalid snapshot data!");
+      vertex.out_edges.reserve(*out_size);
+      for (uint64_t j = 0; j < *out_size; ++j) {
+        auto edge_gid = snapshot.ReadUint();
+        if (!edge_gid) throw RecoveryFailure("Invalid snapshot data!");
+
+        auto to_gid = snapshot.ReadUint();
+        if (!to_gid) throw RecoveryFailure("Invalid snapshot data!");
+        auto edge_type = snapshot.ReadUint();
+        if (!edge_type) throw RecoveryFailure("Invalid snapshot data!");
+
+        auto to_vertex = vertex_acc.find(Gid::FromUint(*to_gid));
+        if (to_vertex == vertex_acc.end()) throw RecoveryFailure("Invalid to vertex!");
+
+        EdgeRef edge_ref(Gid::FromUint(*edge_gid));
+        if (items.properties_on_edges) {
+          // If the snapshot was created with a config where properties are not allowed on edges, then the edges has to
+          // be created here
+          if (snapshot_has_edges) {
+            auto edge = edge_acc.find(Gid::FromUint(*edge_gid));
+            if (edge == edge_acc.end()) throw RecoveryFailure("Invalid edge!");
+            edge_ref = EdgeRef(&*edge);
+          } else {
+            auto [edge, inserted] = edge_acc.insert(Edge{Gid::FromUint(*edge_gid), nullptr});
+            edge_ref = EdgeRef(&*edge);
+          }
+        }
+        vertex.out_edges.emplace_back(get_edge_type_from_id(*edge_type), &*to_vertex, edge_ref);
+        // Increment edge count. We only increment the count here because the
+        // information is duplicated in in_edges.
+        edge_count++;
+      }
+    }
+    ++vertex_it;
+  }
+  spdlog::info("Partial connectivities are recovered.");
+  return {edge_count, highest_edge_gid, first_vertex_gid};
+}
+
+template <typename TFunc>
+void RecoverOnMultipleThreads(const size_t thread_count, const TFunc &func, const std::vector<BatchInfo> &batches) {
+  utils::Synchronized<std::optional<RecoveryFailure>, utils::SpinLock> maybe_error{};
+  {
+    std::atomic<uint64_t> batch_counter = 0;
+    std::vector<std::jthread> threads;
+    threads.reserve(thread_count);
+
+    for (auto i{0U}; i < thread_count; ++i) {
+      threads.emplace_back([&func, &batches, &maybe_error, &batch_counter]() {
+        while (!maybe_error.Lock()->has_value()) {
+          const auto batch_index = batch_counter++;
+          if (batch_index >= batches.size()) {
+            return;
+          }
+          const auto &batch = batches[batch_index];
+          try {
+            func(batch_index, batch);
+          } catch (RecoveryFailure &failure) {
+            *maybe_error.Lock() = std::move(failure);
+          }
+        }
+      });
+    }
+  }
+  if (maybe_error.Lock()->has_value()) {
+    throw RecoveryFailure((*maybe_error.Lock())->what());
```

Comment:
One question about throwing errors. In which case we want to throw error, and in which case do we want to return only error code? This is something that I am not sure, as errors will fill the stack trace, and we could throw error in function which called this function for example, right? What is better to do?

Reply:
In general it is super hard to answer this question. In general I prefer the "rule of thumb" that says use exceptions only when some exceptional thing happens. E.g.:
- `PrepareForWrite` is expected to "fail" when it cannot prepare an object for a transaction to modify => return an error.
- `LoadSnapshot` should only be called with actual snapshot files, therefore it shouldn't fail => throw an exception. 
- And there are a few edge cases too. Should `LoadSnapshot` return an error if it cannot load the snapshot because of some configuration error, such as wrong setting for `properties-on-edges`, or we still consider that as an exceptional case? Etc.

Reply:
Thanks for explanation :smile: Makes sense